python/lsst/daf/butler/registries/sqlPreFlight.py

```diff
@@ -0,0 +1,185 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ("SqlPreFlight")
+
+import itertools
+import copy
+
+from sqlalchemy.sql import select, and_, text
+
+import lsst.log as lsstLog
+
+
+_LOG = lsstLog.Log.getLogger(__name__)
+
+
+class SqlPreFlight:
+    """Class implementing part of preflight solver which extracts
+    units data from registry.
+
+    This is an implementation detail only to be used by SqlRegistry class,
+    not supposed to be used anywhere else.
+
+    Parameters
+    ----------
+    schema : `Schema`
+        Schema instance
+    engine : `sqlalchmey.engine`
+        Engine to use for database access.
+    """
+    def __init__(self, schema, connection):
+        self._schema = schema
+        self._connection = connection
+
+    def selectDataUnits(self, collections, expr, neededDatasetTypes, futureDatasetTypes):
+        r"""Evaluate a filter expression and lists of `DatasetType`\ s and
+        return a set of data unit values.
+
+        Returned set consists of combinations of units participating in data
+        transformation from ``neededDatasetTypes`` to ``futureDatasetTypes``,
+        restricted by existing data and filter expression.
+
+        Parameters
+        ----------
+        collections : `list` of `str`
+            An ordered `list` of collections indicating the Collections to
+            search for Datasets.
+        expr : `str`
+            An expression that limits the `DataUnit`\ s and (indirectly) the
+            Datasets returned.
+        neededDatasetTypes : `list` of `DatasetType`
+            The `list` of `DatasetType`\ s whose instances should be included
+            in the graph and limit its extent.
+        futureDatasetTypes : `list` of `DatasetType`
+            The `list` of `DatasetType`\ s whose instances may be added to the
+            graph later, which requires that their `DataUnit` types must be
+            present in the graph.
```

Comment:
We might want to provide another argument that allows the caller to provide a list of additional DataUnits to be included.  We could even use that to pass the quantum units in for completeness, but I can't think of a case where a SuperTask's quantum units would not be a subset of its DatasetType units.

Reply:
I'm not sure I understand the use case for this but we can certainly extend it later if necessary.